src/interceptors/fetch/index.ts

```diff
@@ -0,0 +1,90 @@
+import {
+  flattenHeadersObject,
+  Headers,
+  headersToObject,
+  objectToHeaders,
+} from 'headers-utils'
+import {
+  Interceptor,
+  IsomoprhicRequest,
+  IsomoprhicResponse,
+  MockedResponse,
+} from '../../createInterceptor'
+
+const debug = require('debug')('fetch')
+
+export const interceptFetch: Interceptor = (observer, resolver) => {
+  const pureFetch = window.fetch
+
+  debug('replacing "window.fetch"...')
+
+  window.fetch = async (input, init) => {
+    const ref = new Request(input, init)
+    const url = typeof input === 'string' ? input : input.url
+    const method = init?.method || 'GET'
+
+    debug('[%s] %s', method, url)
+
+    const request: IsomoprhicRequest = {
+      url: new URL(url, location.origin),
+      method: method,
+      headers: init?.headers ? headersToObject(new Headers(init.headers)) : {},
+      body: await ref.text(),
+    }
+    debug('isomorphic request', request)
+    observer.emit('request', request)
+
+    debug('awaiting for the mocked response...')
+    const response = await resolver(request, ref)
+    debug('mocked response', response)
+
+    if (response) {
+      const isomorphicResponse = normalizeMockedResponse(response)
+      debug('derived isomorphic response', isomorphicResponse)
+
+      observer.emit('response', request, isomorphicResponse)
+
+      return new Response(response.body, {
+        ...isomorphicResponse,
+        // `Response.headers` cannot be instantiated with the `Headers` polyfill.
+        // Apparently, it halts if the `Headers` class contains unknown properties
+        // (i.e. the internal `Headers.map`).
+        headers: flattenHeadersObject(response.headers || {}),
+      })
```

Comment:
Do you think that here we can add a try catch? `new Response` could throw some exceptions for example using a body != null with a status of 204. Maybe we can add a graceful message in this case? 🤔

Reply:
If a new Response instance is given an impossible input, then the exception should surface. That means that the user has provided that impossible input in the mock:

```js
createInterceptor({
  modules: [interceptFetch],
  resolver() {
    return {
      status: 204,
      body: 'hey, i should not be here!'
    }
  }
})
```